vespalib/src/vespa/vespalib/data/smart_buffer.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "smart_buffer.h"
+#include <cassert>
+
+namespace vespalib {
+
+void
+SmartBuffer::ensure_free(size_t bytes)
+{
+    if (write_len() >= bytes) {
+        return;
+    }
+    if ((unused() < bytes) || ((unused() * 3) < read_len())) {
```

Comment:
Curious about the choice of 3 here as 🦄✨magic number✨🦄

Reply:
3, a short story: 3 used to be 4, which seems like a less magic number than 3. (like no-one will question the magic 2 when increasing buffer size). The original thought was probably that we want to compact the buffer if we have at least 25% total free space (1/4). If we have less, we want to expand it as well. The use of 4 was actually an off-by-one error, making the free space threshold 20% instead of 25%. So, in short 3+1 = 4, which is a lot less magic.

We should consider (future additions) having extra options to the smart buffer controlling a soft upper-limit to buffer size (re-size buffer down automatically when it does not need to be too big anymore) as well as space/copy affinity (do we want to copy less or use less memory).